### ISSUE TITLE
Allow --update from a local tarball; allow disabling update warnings

### ISSUE THREAD

**Issue by aldanor** [dbrgn/tealdeer#64]
Two simple changes would make tealdeer more offline-use friendly. In particular, when packaging it for use in an offline environment, one could include tldr's tarball (master.tar.gz) in the package itself, and make it run `tldr --update <tarball>` when installing.

Related: it would be nice to be able to disable "Cache wasn't updated for more than 30 days" warning (maybe at build time) -- if you're in an offline environment, you can't really update it so it's just an annoyance...

**Comment by dbrgn** [comment]
That's a good idea :slightly_smiling_face: 

> Related: it would be nice to be able to disable "Cache wasn't updated for more than 30 days" warning (maybe at build time)

What about an env var?

**Comment by aldanor** [comment]
From packaging perspective, you can’t affect end-user’s environment variables :) You can write a config to the prefix folder though, given that the tool (a) looks for config somewhere in the prefix and (b) this setting is modifiable via the config.

**Comment by dbrgn** [comment]
> From packaging perspective, you can’t affect end-user’s environment variables :)

Yes, but a user can disable the warning on a system where no network connectivity is available by setting that env var.

We already have config file support though, I kind of forgot about that. We could put the option there.

**Comment by aldanor** [comment]
> Yes, but a user can disable the warning on a system where no network connectivity is available by setting that env var.

Technically, yes, I agree :) In my case I'm packaging it for an environment where it's 100% that end users would have no connectivity... in which case it would a bit uncool to have to tell all of them "please don't forget to set this cryptic environment variable to get rid of warnings".

**Comment by aldanor** [comment]
> We already have config file support though, I kind of forgot about that. We could put the option there.

Would any location inside the prefix be looked up for the config? (i.e., if it's installed to `$PREFIX/bin/tldr`)

**Comment by dbrgn** [comment]
Now I'm curious, what kind of system is that? :slightly_smiling_face: 

> Would any location inside the prefix be looked up for the config?

Right now we use XDG to determine the proper location for the config file, once #59 is merged it will be the [appdirs](https://github.com/djc/appdirs-rs) crate. See https://github.com/Bassetts/tealdeer/blob/3dafe92f53376bef3cd7ca4d3952fa01857d9904/src/config.rs#L187-L208.

On Linux, that will be `$XDG_CONFIG_HOME/tealdeer/config.toml` (usually `$XDG_CONFIG_HOME` is `~/.config/`).

You can also override it using the `TEALDEER_CONFIG_DIR` env var.

> In which case it would a bit uncool to have to tell all of them "please don't forget to set this cryptic environment variable to get rid of warnings".

I'm open to adding a compile flag that would disable networking support. Maybe we could then even get rid of the HTTP client libraries for that build target?

**Comment by aldanor** [comment]
> Now I'm curious, what kind of system is that?

Corporate environment :dollar: :smile: 

> Right now we use XDG to determine the proper location for the config file, once #59 is merged it will be the appdirs crate. See https://github.com/Bassetts/tealdeer/blob/3dafe92f53376bef3cd7ca4d3952fa01857d9904/src/config.rs#L187-L208.
>
> On Linux, that will be $XDG_CONFIG_HOME/tealdeer/config.toml (usually $XDG_CONFIG_HOME is ~/.config/).

From packaging perspective, you can't touch the user's home, only the prefix :)

> You can also override it using the TEALDEER_CONFIG_DIR env var.

From packaging perspective, you can't, only the end-user can :)

> I'm open to adding a compile flag that would disable networking support. Maybe we could then even get rid of the HTTP client libraries for that build target?

That's actually a fair point. `reqwest` alone has 18 dependencies, along which are huge crates like `tokio`, `hyper`, each with a dozen more deps. You could probably trim down the `tldr` app substantially were you to drop networking support.

**Comment by dbrgn** [comment]
I started implementing this, should already be mostly done. However, I hit a few issues with the feature flags while building integration tests. It will probably have to wait a few days.

**Comment by aldanor** [comment]
Thanks for a prompt response, I’d be sure happy to help test this :)

// Slightly out of scope of this (but related), so I don’t forget - the current reqwest-based update doesn’t work behind a corporate proxy, maybe because it’s messing with certificates. I can curl/wget things from command line just fine, cargo also works without problems, but not tldr update. Will try to roll back to a previous curl-based version tomorrow and report back.

**Comment by aldanor** [comment]
> I started implementing this, should already be mostly done. However, I hit a few issues with the feature flags while building integration tests. It will probably have to wait a few days.

Any chances for this? :)

**Comment by dbrgn** [comment]
@aldanor see #72 :slightly_smiling_face:

**Comment by dbrgn** [comment]
Closing due to lack of interest.
